Blockchain: refactor: Add abstract DB operation, let DB manage the implementation of this OP + make DB handle Cache

jochem-brouwer opened the pull request:
Blockchains `index.ts` used to do a lot of DB-like operations, like calculating the keys and updating the cache. This is now different; a `databaseOperation` factory is added, which describes these operations. Also, all DB operations are moved into a `DB` folder. This is an internal refactor and does not change any external behavior. Cherry-picked from #895.

ryanio reviewed (approved):
LGTM, looks really clean!

jochem-brouwer commented:
I'd like an additional review here before merging this, since it changes the internal stuff a lot, rather have some more eyes on it 😄 

@holgerd77 @cgewecke @rumkin

cgewecke commented on packages/blockchain/src/db/dbManager.ts:
```diff
@@ -129,84 +111,84 @@ export class DBManager {
+  async batch(ops: DatabaseOperation[]) {
+    const convertedOps: DBOp[] = ops.map((op) => op.baseDBOp)
+    // update the current cache for each operation
+    ops.map((op) => op.updateCache(this._cache))
```
Is this a substantive change (a bug fix)? (Am just requesting more info - I don't know this lib very well)

jochem-brouwer replied:
This PR does not fix any bugs. Previously, these cache operations were all handled in `index.ts` of blockchain, but now they are handled within these `DatabaseOperation`s every time we do a batch request of these `ops`.

cgewecke commented on packages/blockchain/src/db/databaseOperation.ts:
```diff
@@ -0,0 +1,145 @@
+import BN = require('bn.js')
```
Could this be?
```js
import { BN } from 'ethereumjs-util'
```

jochem-brouwer replied:
Yes, will change!

cgewecke replied:
07c5f70

cgewecke reviewed (dismissed):
Just to echo @ryanio's review this looks great! Really nice. 🎉 

Couple questions, that's it.

cgewecke reviewed (approved):
Re-approving 🙂

jochem-brouwer commented:
I will wait until @holgerd77 also gives this a look (don't worry, not in a hurry). Have to rebase this anyways, so that will need a re-approval anyways 😛

holgerd77 commented:
Rebased this.

holgerd77 commented:
I would perceive this as cleaner and less redundant if we remove the `DB` (or similar) name part from the `db` folder file names, so:

- `db/databaseOperation.ts` -> `db/operation.ts`
- `db/dbConstants` -> `db/constants.ts`
- `db/dbManager` -> `db/manager.ts`

Prefixes on classes (e.g. `class DBManager`) or other code structures shouldn't be removed I would say to keep code readability on imports, but we should unify the prefix here as well, there is currently a mixture of `DB` and `Database` (`DBManager`, `DatabaseOperation`,...). I would have a strong tendency to the shorter `DB` prefix, so e.g.:

- `DatabaseOperation` -> `DBOperation`
- `DatabaseOperationTarget` -> `DBOperationTarget`

And so on.

holgerd77 commented:
Yeah, after letting the code take a bit its effect I would re-state that I find this renaming thing pretty important, this otherwise leads to very long lines which are somewhat hard to read.

```typescript
ops.push(DatabaseOperation.del(DatabaseOperationTarget.Header, { blockHash, blockNumber }))
```

```typescript
ops.push(DBOps.del(DBTarget.Header, { blockHash, blockNumber }))
```

Just doing a live-test here on myself with the two lines from above to see how I perceive readability. I think I would actually also tend to do the `DatabaseOperationTarget` -> `DBTarget` renaming, think there is not lost anything since these targets very much speak for themselves and this is a lot more comprise.

Apart from these cosmetics this looks really good! Great that this library gets such a shift on readability! 😄

jochem-brouwer commented:
Hi @holgerd77, thanks for your suggestions! You are absolutely right about the names of these files and the class names, I will edit those. You are also right about the names of these newly introduced helpers, I honestly had a struggle myself to come up with a name for them which was both short and descriptive 😅 .

jochem-brouwer commented:
OK, in d747941 I have renamed DatabaseOperation -> DBOp, the original exported `DBOp` type (which is used in `level`) to `DBOperationData` and `DatabaseOperationTarget` to `DBTarget`. Also renamed the files in the `db` directory as suggested.

Ready for re-review! @holgerd77 

(I thought I pulled and then rebased, but looks like I did not pull your rebase changes?)

holgerd77 commented:
Rebased this on top of #929

holgerd77 reviewed (approved):
Have done some more renamings, this looks great, thanks Jochem! 😄  👍 

Tx test CI run was failing on last run, tests itself are passing though, this is just a browser timing issue and also highly unlikely to be related with the changes here, since tx has no block or blockchain dependencies.

We might want to observe this a bit, but I will merge here for now.

jochem-brouwer commented:
Thanks for these extra renames 👍 

Tx and Block sometimes indeed seem to timeout; if you rerun the CI then it usually passes.
